src/openzeppelin/tests/test_accesscontrol.cairo

```diff
@@ -2,124 +2,308 @@ use openzeppelin::access::accesscontrol::AccessControl;
+
+fn OTHER_ADMIN() -> ContractAddress {
+    contract_address_const::<4>()
+}
+
 fn setup() {
-    set_caller_address(ACCOUNT1());
-    AccessControl::constructor(ACCOUNT1());
+    AccessControl::_grant_role(DEFAULT_ADMIN_ROLE, ADMIN());
+    set_caller_address(ADMIN());
 }
 
+//
+// initializer
+//
+
 #[test]
 #[available_gas(2000000)]
 fn test_initializer() {
     AccessControl::initializer();
     assert(AccessControl::supports_interface(IACCESSCONTROL_ID), 'Should support own interface');
 }
 
+//
+// has_role
+//
+
 #[test]
 #[available_gas(2000000)]
-fn test_constructor() {
-    AccessControl::constructor(ACCOUNT1());
-    assert(AccessControl::has_role(DEFAULT_ADMIN_ROLE, ACCOUNT1()), 'Admin role should be set');
-    assert(AccessControl::supports_interface(IACCESSCONTROL_ID), 'Should support own interface');
+fn test_has_role() {
+    setup();
+    assert(!AccessControl::has_role(ROLE, AUTHORIZED()), 'should not have role');
+    AccessControl::_grant_role(ROLE, AUTHORIZED());
+    assert(AccessControl::has_role(ROLE, AUTHORIZED()), 'should have role');
 }
 
+
+//
+// assert_only_role
+//
+
 #[test]
 #[available_gas(2000000)]
 fn test_assert_only_role() {
     setup();
-    AccessControl::assert_only_role(DEFAULT_ADMIN_ROLE);
+    AccessControl::grant_role(ROLE, AUTHORIZED());
+    set_caller_address(AUTHORIZED());
+    AccessControl::assert_only_role(ROLE);
 }
 
 #[test]
 #[available_gas(2000000)]
 #[should_panic(expected: ('Caller is missing role', ))]
-fn test_assert_only_role_unauthorized() {
+fn test_assert_only_role_unauthorized_1() {
     setup();
-    set_caller_address(ACCOUNT2());
-    AccessControl::assert_only_role(DEFAULT_ADMIN_ROLE);
+    set_caller_address(OTHER());
+    AccessControl::assert_only_role(ROLE);
 }
 
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected: ('Caller is missing role', ))]
+fn test_assert_only_role_unauthorized_2() {
+    setup();
+    set_caller_address(AUTHORIZED());
+    AccessControl::assert_only_role(OTHER_ROLE);
+}
+
+//
+// grant_role
+//
+
 #[test]
 #[available_gas(2000000)]
 fn test_grant_role() {
     setup();
-    AccessControl::grant_role(DEFAULT_ADMIN_ROLE, ACCOUNT2());
-    assert(AccessControl::has_role(DEFAULT_ADMIN_ROLE, ACCOUNT2()), 'Role should be granted');
+    AccessControl::grant_role(ROLE, AUTHORIZED());
+    assert(AccessControl::has_role(ROLE, AUTHORIZED()), 'Role should be granted');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_grant_role_multiple_times_for_granted_role() {
+    setup();
+    AccessControl::grant_role(ROLE, AUTHORIZED());
+    AccessControl::grant_role(ROLE, AUTHORIZED());
+    assert(AccessControl::has_role(ROLE, AUTHORIZED()), 'Role should still be granted');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_grant_role_already_granted() {
+    setup();
+    AccessControl::grant_role(ROLE, AUTHORIZED());
+    AccessControl::grant_role(ROLE, AUTHORIZED());
+    assert(AccessControl::has_role(ROLE, AUTHORIZED()), 'Role should still be held');
```

Comment:
how is this test different from the one above? i think i get the difference is that the above intends to check that it doesn't revert/passes, while the one below intends to check that the account still has the role. but are they really different?

Reply:
Haha yeah, you're right. One should work